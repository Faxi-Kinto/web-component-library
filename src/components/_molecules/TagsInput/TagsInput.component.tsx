import React, { useCallback, useMemo, useRef, useState } from 'react';
import classNames from 'classnames';
import uniqid from 'uniqid';

import * as Styled from './TagsInput.styles';

type ValidateFn = (value: string) => React.ReactNode | undefined;

type Tag = {
  value: string;
  error: React.ReactNode;
};

type TagsState = Record<string, Tag>;

const removeInvalidTags = (tags: string | string[] = []) => {
  const refinedTags = typeof tags === 'string' ? tags.split(',') : tags;
  return Array.from(new Set(refinedTags.filter(Boolean)));
};

const validateTag = (tag: string, validate: ValidateFn | ValidateFn[]) => {
  if (Array.isArray(validate)) {
    for (const validation of validate) {
      const result = validation(tag);
      if (result) return result;
    }
  } else {
    return validate(tag);
  }
};

const createTagsFromStrings = (
  strings: string[],
  validate?: ValidateFn | ValidateFn[]
): TagsState =>
  removeInvalidTags(strings).reduce((acc, tag) => {
    const key = uniqid();

    acc[key] = {
      value: tag,
      error: validate ? validateTag(tag, validate) : '',
    };
    return acc;
  }, {} as TagsState);

type TagsInputProps = {
  placeholder: string;
  className?: string;
  label?: string;
  initialTags?: string[];
  spaceSeparates?: boolean;
  validateTag?: ValidateFn | ValidateFn[];
  onChange?: (tags: string[]) => void;
};

const TagsInput: React.FC<TagsInputProps> = (
  props: TagsInputProps
): JSX.Element => {
  const {
    initialTags,
    className,
    label,
    spaceSeparates = true,
    placeholder = 'Add email',
    validateTag,
    onChange,
  } = props;

  const [tags, setTags] = useState<TagsState>(
    initialTags ? createTagsFromStrings(initialTags, validateTag) : {}
  );

  const [inputValue, setInputValue] = useState('');

  const inputRef = useRef<HTMLInputElement>(null);

  const addTag = useCallback(
    (value: string) => {
      setTags(old => {
        if (
          Object.values(old)
            .map(({ value: v }) => v)
            .includes(value)
        )
          return old;
        else {
          const newTags = {
            ...old,
            ...createTagsFromStrings([value], validateTag),
          };
          requestAnimationFrame(() =>
            onChange?.(Object.values(newTags).map(tag => tag.value))
          );

          return newTags;
        }
      });
      setInputValue('');
    },
    [validateTag, onChange]
  );

  const removeTag = useCallback(
    (key?: string) => {
      setTags(old => {
        const newTags = { ...old };
        if (key) {
          delete newTags[key];
          onChange?.(Object.values(newTags).map(tag => tag.value));
          return newTags;
        }
        const tagKeys = Object.keys(old);
        delete newTags[tagKeys[tagKeys.length - 1]];
        requestAnimationFrame(() =>
          onChange?.(Object.values(newTags).map(tag => tag.value))
        );
        return newTags;
      });
    },
    [onChange]
  );

  const onKeyDown = useCallback(
    (ev: React.KeyboardEvent) => {
      const { key } = ev;

      if (key === 'Enter' || (spaceSeparates && key === ' ')) {
        ev.preventDefault();
        const trimmedVal = inputValue.trim();
        if (trimmedVal) addTag(trimmedVal);
      }

      if (key === 'Backspace') {
        if (!inputValue) {
          removeTag();
        } else if (!inputValue.trim()) {
          setInputValue('');
        }
      }
    },
    [addTag, removeTag, inputValue, spaceSeparates]
  );

  const tagError = useMemo(
    () => Object.entries(tags).find(entry => Boolean(entry[1].error)),
    [tags]
  );

  return (
    <Styled.Container className={className}>
      {label && (
        <label
          className={classNames({ 'tags-label--error': !!tagError })}
          onClick={() => inputRef.current?.focus()}
        >
          {label}
        </label>
      )}
      {tagError && <div className="tags-error">{tagError?.[1].error}</div>}
      <div className="tags-input" onClick={() => inputRef.current?.focus()}>
        <div className="tags-input__wrapper">
          {Object.entries(tags).map(([key, { value }]) => (
            <div
              className={classNames([
                'tags-input__entry',
                { 'tags-input__entry--error': tagError?.[0] === key },
              ])}
              key={key}
            >
              <span>{value}</span>
              <button type="button" onClick={() => removeTag(key)}>
                <i className="close" />
              </button>
            </div>
          ))}
          <input
            ref={inputRef}
            value={inputValue}
            onChange={ev => setInputValue(ev.target.value)}
            onBlur={ev => addTag(ev.target.value.trim())}
            onKeyDown={onKeyDown}
          />

          {!Object.keys(tags).length && !inputValue && (
            <span className="tags-input__placeholder">{placeholder}</span>
          )}
        </div>
      </div>
    </Styled.Container>
  );
};

export default TagsInput;
