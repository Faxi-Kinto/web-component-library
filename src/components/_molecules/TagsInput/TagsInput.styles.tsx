import {
  flexRow,
  fontSize,
  positionAbsolute,
  pxToRem,
  size,
} from '@faxi/web-css-utilities';
import styled from 'styled-components';

export const Container = styled.div`
  label,
  .tags-error {
    ${fontSize('14px', '20px')};

    font-weight: 500;
  }

  label {
    display: block;
    margin-bottom: ${pxToRem('24px')};
  }

  .tags-label {
    &--error {
      margin-bottom: 0;
    }
  }

  .tags-error {
    color: #eb4820;
    margin-top: ${pxToRem('4px')};
  }

  .tags-input {
    border-bottom: solid 1px #95a1a6;

    &__wrapper {
      ${fontSize('16px', '23px')};

      margin: 0 -${pxToRem('4px')};
      position: relative;
      padding-bottom: ${pxToRem('12px')};
      min-height: ${pxToRem('25px')};
      height: auto;
      box-sizing: content-box;
      display: flex;
      align-items: flex-start;
      flex-wrap: wrap;

      > * {
        margin: ${pxToRem('4px')};
      }

      input {
        ${size('140px', '25px')};

        font: inherit;
        color: #95a1a6;
        border: none;

        &:focus {
          outline: none;

          & + .tags-input__placeholder {
            display: none;
          }
        }
      }
    }

    &__entry {
      ${flexRow('center', 'center')};

      min-height: ${pxToRem('25px')};
      padding: 0 ${pxToRem('8px')};
      color: #95a1a6;
      border: solid 1px;
      word-break: break-word;

      &--error {
        border-color: #eb4820;
      }

      > button {
        ${size('20px')};

        cursor: pointer;
        display: flex;
        align-items: center;
        padding: 0;
        background: none;
        margin-left: 0.5rem;
        border: none;

        &:focus {
          outline: none;
        }

        i.close {
          ${size('100%')};

          display: inline-block;
          position: relative;

          &::before,
          &::after {
            ${size('12px', '2px')};
            ${positionAbsolute('50%', '50%')};

            content: '';
            background-color: #10a5b5;
          }

          &::before {
            transform: translate(50%, -50%) rotateZ(45deg);
          }

          &::after {
            transform: translate(50%, -50%) rotateZ(-45deg);
          }
        }
      }
    }

    &__placeholder {
      position: absolute;
      color: #95a1a6;
      opacity: 0.5;
    }
  }
`;
