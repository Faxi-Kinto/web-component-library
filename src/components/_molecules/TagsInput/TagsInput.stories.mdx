import TagsInput from './';
import { action } from '@storybook/addon-actions';

<Meta title="Core|Molecules/TagsInput" />

# TagsInput Component

## Implementation

### Primary

<Preview>
  <Story name="primary">
    <TagsInput label="Email" onChange={action('tags-change')} />
  </Story>
</Preview>

### Prefilled

<Preview>
  <Story name="prefilled">
    <TagsInput
      label="Email"
      initialTags={['some@email.com', 'test']}
    />
  </Story>
</Preview>

### With validation

<Preview>
  <Story name="with validation">
    <TagsInput
      label="Email"
      validateTag={tag => (tag.includes('@') ? '' : 'Invalid email')}
    />
  </Story>
</Preview>

## Props

<Props of={TagsInput} />

### Prefilled with validaiton

<Preview>
  <Story name="prefilled with validation">
    <TagsInput
      label="Email"
      initialTags={['some@email.com', 'test']}
      validateTag={tag => (tag.includes('@') ? '' : 'Invalid email')}
    />
  </Story>
</Preview>
