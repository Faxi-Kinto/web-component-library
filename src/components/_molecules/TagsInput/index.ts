import TagsInput from './TagsInput.component';

export default TagsInput;
