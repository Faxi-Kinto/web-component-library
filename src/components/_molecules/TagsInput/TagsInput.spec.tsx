import React from 'react';
import renderer from 'react-test-renderer';
import TagsInput from './TagsInput.component';

test('TagInput renders correctly', () => {
  const component = renderer.create(<TagsInput placeholder="tags" />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
