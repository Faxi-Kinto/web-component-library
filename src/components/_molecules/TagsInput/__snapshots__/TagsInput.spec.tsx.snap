// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagInput renders correctly 1`] = `
.c0 label,
.c0 .tags-error {
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 500;
}

.c0 label {
  display: block;
  margin-bottom: 1.5rem;
}

.c0 .tags-label--error {
  margin-bottom: 0;
}

.c0 .tags-error {
  color: #eb4820;
  margin-top: 0.25rem;
}

.c0 .tags-input {
  border-bottom: solid 1px #95a1a6;
}

.c0 .tags-input__wrapper {
  font-size: 1rem;
  line-height: 1.4375rem;
  margin: 0 -0.25rem;
  position: relative;
  padding-bottom: 0.75rem;
  min-height: 1.5625rem;
  height: auto;
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 .tags-input__wrapper > * {
  margin: 0.25rem;
}

.c0 .tags-input__wrapper input {
  width: 140px;
  height: 25px;
  font: inherit;
  color: #95a1a6;
  border: none;
}

.c0 .tags-input__wrapper input:focus {
  outline: none;
}

.c0 .tags-input__wrapper input:focus + .tags-input__placeholder {
  display: none;
}

.c0 .tags-input__entry {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 1.5625rem;
  padding: 0 0.5rem;
  color: #95a1a6;
  border: solid 1px;
  word-break: break-word;
}

.c0 .tags-input__entry--error {
  border-color: #eb4820;
}

.c0 .tags-input__entry > button {
  width: 20px;
  height: 20px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  background: none;
  margin-left: 0.5rem;
  border: none;
}

.c0 .tags-input__entry > button:focus {
  outline: none;
}

.c0 .tags-input__entry > button i.close {
  width: 100%;
  height: 100%;
  display: inline-block;
  position: relative;
}

.c0 .tags-input__entry > button i.close::before,
.c0 .tags-input__entry > button i.close::after {
  width: 12px;
  height: 2px;
  position: absolute;
  top: 50%;
  right: 50%;
  content: '';
  background-color: #10a5b5;
}

.c0 .tags-input__entry > button i.close::before {
  -webkit-transform: translate(50%,-50%) rotateZ(45deg);
  -ms-transform: translate(50%,-50%) rotateZ(45deg);
  transform: translate(50%,-50%) rotateZ(45deg);
}

.c0 .tags-input__entry > button i.close::after {
  -webkit-transform: translate(50%,-50%) rotateZ(-45deg);
  -ms-transform: translate(50%,-50%) rotateZ(-45deg);
  transform: translate(50%,-50%) rotateZ(-45deg);
}

.c0 .tags-input__placeholder {
  position: absolute;
  color: #95a1a6;
  opacity: 0.5;
}

<div
  className="c0"
>
  <div
    className="tags-input"
    onClick={[Function]}
  >
    <div
      className="tags-input__wrapper"
    >
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        value=""
      />
      <span
        className="tags-input__placeholder"
      >
        tags
      </span>
    </div>
  </div>
</div>
`;
